pkg/agent/controller/l7visibility/l7visibility_controller_test.go

```diff
@@ -0,0 +1,546 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   namespace,
+			Annotations: annotations,
+		},
+		Spec: v1.PodSpec{
+			NodeName: "fakeNode",
+		},
+		Status: v1.PodStatus{
+			PodIP: "10.0.0.1",
+		},
+	}
+}
+
+func newNamespaceObject(name string, annotations map[string]string) *v1.Namespace {
+	return &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Annotations: annotations,
+			Labels: map[string]string{
+				"fakeKey": "fakeValue",
+			},
+		},
+	}
+}
+
+func newPodInterface(podName, podNamespace string, ofPort int32) *interfacestore.InterfaceConfig {
+	containerName := k8s.NamespacedName(podNamespace, podName)
+	return &interfacestore.InterfaceConfig{
+		InterfaceName:            util.GenerateContainerInterfaceName(podName, podNamespace, containerName),
+		ContainerInterfaceConfig: &interfacestore.ContainerInterfaceConfig{PodName: podName, PodNamespace: podNamespace, ContainerID: containerName},
+		OVSPortConfig:            &interfacestore.OVSPortConfig{OFPort: ofPort},
+	}
+}
+
+func waitEvents(t *testing.T, expectedEvents int, c *fakeController) {
+	require.NoError(t, wait.Poll(100*time.Millisecond, 5*time.Second, func() (done bool, err error) {
+		return c.queue.Len() == expectedEvents, nil
+	}))
+}
+
+func TestPodAdd(t *testing.T) {
+	var targetPort uint32
+	testNS1 := newNamespaceObject("test-ns1", annotationsEmpty)
+	pod1 := newPodObject("test-pod1", "test-ns1", annotationsCorrectIngress)
+	pod2 := newPodObject("test-pod2", "test-ns1", annotationsIncorrect)
+	pod1NN := k8s.NamespacedName(pod1.Namespace, pod1.Name)
+	podInterface1 := newPodInterface(pod1.Name, pod1.Namespace, int32(pod1OFPort))
+	podInterface2 := newPodInterface(pod2.Name, pod2.Namespace, int32(pod2OFPort))
+	interfaces := []*interfacestore.InterfaceConfig{
+		podInterface1,
+		podInterface2,
+	}
+	testcases := []struct {
+		name                      string
+		addedPod                  *v1.Pod
+		expectedPodToDirectionMap map[string]v1alpha2.Direction
+		expectedCalls             func(mockOFClient *openflowtest.MockClient)
+		expectedError             string
+	}{
+		{
+			name:     "Add pod with correct annotations",
+			addedPod: pod1,
+			expectedPodToDirectionMap: map[string]v1alpha2.Direction{
+				pod1NN: v1alpha2.DirectionIngress,
+			},
+			expectedCalls: func(mockOFClient *openflowtest.MockClient) {
+				mockOFClient.EXPECT().InstallTrafficControlMarkFlows(fmt.Sprintf("tcl7:%s", pod1NN), []uint32{uint32(podInterface1.OFPort)}, targetPort, directionIngress, actionMirror, types.TrafficControlFlowPriorityLow)
+			},
+		}, {
+			name:                      "Add pod with incorrect annotations",
+			addedPod:                  pod2,
+			expectedPodToDirectionMap: map[string]v1alpha2.Direction{},
+			expectedCalls:             func(mockOFClient *openflowtest.MockClient) {},
+			expectedError:             fmt.Sprintf("annotation key %s can only have values (Ingress/Egress/Both)", types.L7VisibilityAnnotationKey),
```

Comment:
I would just use `errInvalidAnnotation` here instead of repeating the error string